### ISSUE TITLE
[improvement] Type resolution/handling

### ISSUE THREAD

**Issue by elythyr** [phpactor/phpactor#865]
After working to handle typed properties and nullable types.
I realize that we are doing it wrong in a lot of places

For instance: https://github.com/phpactor/code-transform/blob/6e74b266ec85b68dcbcd620631b7b85f1a742322/lib/Adapter/WorseReflection/Refactor/WorseExtractMethod.php#L210
Here we are building the parameters for the method to extract.
We can see `$parameterBuilder->type($variableType->short());` to define the type of the parameter from a string.
- `short()` return the last piece of a FQCN, no handling of partial or fully qualified names
- we lost any piece of information about the type, the nullable aspect for instance

Could we consider gaving the `Type` object directly to the consumer ?
I guess it wasn't done to keep the coupling as low as possible, but I would argue that the `WorseReflection` is generic enough so that we can afford it since it's already build around Adapters, we mainly use interfaces of facade.

We also got `$builder->use((string) $variableType);` to add a use statement if not already present.
I will pass on the fact that we might not want to do that here, it's another topic.
And focus on the fact that with the string representation of a Type include the leading `?` if nullable, we could think about removing it, but I don't mind it in a `__toString` method.
We should only be able to import two things if I'm correct: FQCN and functions.
Since we don't handle use clause for function yet we can ignore that.
And pass the `Type` once again, in order to not loose any information. 

To go back quickly on the `short()` issue, I think we should come with a way to retrieve the type by considering the current namespace and the import table.
I don't know if it would be best to have a dedicated util class for that or any thing else.
But there is I think something to do to improve that.

**Comment by dantleech** [comment]
Yes this is a long standing issue (there's a previous one by @wierdan which I can't find from my mobile). We need indeed to preserve the type infornation.

**Comment by elythyr** [comment]
This is my proposal:
- For the `Type` in `worse-reflection`
  - `Type::__toString()`
    - Always return the complete type (FQCN)
    - Include the `?` for nullable types (for me it's part of the type)
  - `Type::short()`
    - Always return the shortest version of a type (not even sure we will still having a need for it)
    - Must not include the `?`
  - `Type::primitive()`
    - Must return null if `Type::isPrimitive()` returns false
    - Must not include the `?`
    - Could be replaced by `Type::short()` and then remove so we don't deal with `null` values
- For the `Type` in `code-builder`
  - The class should be considered as an adapter for the `Type` from `worse-reflection`
    - Meaning it should accept it as parameter in it's constructor and delegate the construction from a `string` to it
  - `Type::notNone()` & `Type::$none` are not used anymore and should be removed
    - We can deprecate the method and make it always return false for now
- Add a `TypeResolver` class in `code-builder`
  - Have the responsibility to resolve a type given a scope so that partially qualified names can be handled
  - Add an extension for Twig ?
  - How to deal with multiple scopes ? For example two namespaces in the same file...

**Comment by dantleech** [comment]
We introduced:

https://github.com/phpactor/worse-reflection/blob/93e987f3d55de5040169b5b9eca395c036726b97/lib/Core/Reflection/ReflectionScope.php#L12

Example:

https://github.com/phpactor/code-transform/blob/master/lib/Adapter/WorseReflection/Transformer/CompleteConstructor.php#L74

Which would solve this issue I think if I applied here.
